dlt/extract/incremental.py

```diff
@@ -0,0 +1,174 @@
+    new records created since the last run of the pipeline.
+
+    To use this the resource function should have an argument either type annotated with `Incremental` or a default `Incremental` instance.
+    For example:
+
+    >>> @dlt.resource(primary_key='id')
+    >>> def some_data(created_at=dlt.Incremental('created_at', '2023-01-01T00:00:00Z'):
+    >>>    yield from request_data(created_after=created_at.last_value)
+
+    When the resource has a `primary_key` specified this is used to deduplicate overlapping items with the same cursor value.
+
+    Args:
+        cursor_column: The name of the cursor column. This can be a column name or any valid JSON path.
+        initial_value: Optional value used for `last_value` when no state is available, e.g. on the first run of the pipeline. If not provided `last_value` will be `None` on the first run.
+        last_value_func: Callable used to determine which cursor value to save in state. It is called with the stored state value and all cursor vals from currently processing items. Default is `max`
+    """
+    def __init__(
+            self,
+            cursor_column: str,
+            initial_value: Optional[TCursorValue]=None,
+            last_value_func: Optional[LastValueFunc]=None,
+    ) -> None:
+        assert cursor_column, "`cursor_column` must be a column name or json path"
+        self.cursor_column = cursor_column
+        self.cursor_column_p = JSONPath(cursor_column)
+        self.initial_value = initial_value
+        self.last_value_func = last_value_func or max
+        from dlt.pipeline.current import resource_state  # TODO: Resolve circular import
+        self._state = resource_state
+
+    def copy(self) -> "Incremental[TCursorValue, TUniqueValue]":
+        return self.__class__(self.cursor_column, initial_value=self.initial_value, last_value_func=self.last_value_func)
+
+    @classmethod
+    def from_config(cls, cfg: IncrementalConfigSpec) -> "Incremental[Any, Any]":
+        # TODO: last_value_func from name
+        return cls(cfg.cursor_column, initial_value=cfg.initial_value)
+
+    @property
+    def state(self) -> IncrementalColumnState:
+        return self._state().setdefault('incremental', {}).setdefault(self.cursor_column, {'last_value': json.loads(json.dumps(self.initial_value)), 'unique_keys': []}) # type: ignore
+
+    @property
+    def last_value(self) -> Optional[TCursorValue]:
+        return self.state['last_value']  # type: ignore
+
+    @property
+    def unique_keys(self) -> List[TUniqueValue]:
+        return self.state['unique_keys']
+
+    def __call__(self, row: TDataItem, primary_key: Optional[TTableHintTemplate[TColumnKey]]) -> bool:
+        if row is None:
+            return True
+        state = self.state
+        last_value = state['last_value']
+        row_value = json.loads(json.dumps(self.cursor_column_p.parse(row)[0]))  # For now the value needs to match deserialized presentation from state
```

Comment:
I get the problem. What about making our `last_value_func` always working on string representations?
- all types of interest (int, date, timestamp, decimal, float(?)) are also string comparable
- then you can just use `serialize_value` from `configuration/utils`
- you can always write a custom last value func and use `deserialize_value` to go back to the native representation

Reply:
String compare like this? :grimacing:  `max("11", "100")` 

last_value_func is not always called i.e. `if last_value == row_value:`  and on the first row with clean state. But now thinking it should always be used anyway. You may have some special parsing in a custom function.

Reply:
I think best would be to use the typed json dumper for state. Least surprising when you get the same type back as you put in.

Reply:
my idea with comparing integers as string was amazing 😁